File: cpp/0216_combinationSum3.cpp
// 1.回溯范围：所有1-9的组合
// 2.剪枝条件：当数量超过k或者相加之和大于n
// 3.终止：当数量等于k且相加之和等于n
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(res,k,n,comb,1);
        return res;
    }
    void backtrack(vector<vector<int>> &res, int k, int target, vector<int> &comb, int pos)
    {
        if(comb.size()==k && target==0)
        {
            res.push_back(comb);
            return;
        }
        for(int i=pos;i<10;i++)//1-9
        {
            if(target-i<0 || comb.size()>k)
                return;
            comb.push_back(i);
            backtrack(res,k,target-i,comb,i+1);
            comb.pop_back();
        }
    }
};

File: java/0216_combinationSum3.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
        backtrack(res,comb,k,n,1);
        return res;
    }
    public void backtrack(List<List<Integer>> res, List<Integer> comb, int k, int target, int pos)
    {
        if(target==0 && comb.size()==k)
        {
            res.add(new ArrayList<Integer>(comb));
            return;
        }
        for(var i=pos;i<10;i++)
        {
            if(target-i<0 || comb.size()>k)
                return;
            comb.add(i);
            backtrack(res,comb,k,target-i,i+1);
            comb.remove(comb.size()-1);
        }
    }
}
